### PR TITLE
NSQ Support for Scala Stream Collector and Stream Enrich

### PR DESCRIPTION
**Pull request by aldemirenes**

### CODE REVIEW

File: 3-enrich/stream-enrich/project/BuildSettings.scala [hunk 1]
```diff
@@ -46,11 +46,9 @@ object BuildSettings {
...
-    // Executable jarfile
-    assemblyOption in assembly ~= { _.copy(prependShellScript = Some(defaultShellScript)) },
     // Name it as an executable
-    jarName in assembly := { s"${name.value}-${version.value}" }
+    jarName in assembly := { s"${name.value}-${version.value}.jar" }
   )
 
   lazy val buildSettings = basicSettings ++ scalifySettings ++ sbtAssemblySettings
-}
+}
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/project/BuildSettings.scala]: those changes will be taken care of by separate tickets: e.g. #3311

File: 3-enrich/stream-enrich/project/Dependencies.scala [hunk 2]
```diff
@@ -36,20 +36,21 @@ object Dependencies {
...
-    val igluClient           = "0.4.0"
+    val scalaz7              = "7.0.9"
+    val igluClient           = "0.5.0"
     val snowplowTracker      = "0.3.0"
     // Scala (test only)
     // val specs2               = "2.3.6" Conflicts with com.chuusai:shapeless
-    val specs2               = "2.2.3"
-    val scalazSpecs2         = "0.1.2"
+    val specs2               = "2.3.11"
+    val scalazSpecs2         = "0.2"
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/project/Dependencies.scala]: please try to keep the dependency changes to a minimum, they will be taken care of somewhere else

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+// Snowplow
+import com.snowplowanalytics.snowplow.collectors.thrift._
+
+/**
+  * Stdouterr Sink for Scala enrichment
+  */
+class NSQSink(config: KinesisEnrichConfig, inputType: InputType.InputType) extends ISink {
+
+  val producer = new NSQProducer().addAddress("localhost", config.producerPort).start();
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: same shouldn't the host be part of the configuration?

Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: and what about closing?

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.kinesis
+package sinks
+
+// NSQ
+import com.github.brainlag.nsq.NSQProducer
+
+// Snowplow
+import com.snowplowanalytics.snowplow.collectors.thrift._
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: I don't think this is used

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    * to the given output stream.
+    *
+    * EnrichedEvent takes the form of a tab-delimited
+    * String until such time as https://github.com/snowplow/snowplow/issues/211
+    * is implemented.
+    *
+    * @param events Sequence of enriched events and (unused) partition keys
+    * @return Whether to checkpoint
+    */
+
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: the scaladoc doesn't have the same indent than the method.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  def storeEnrichedEvents(events: List[(String, String)]): Boolean = {
+    inputType match {
+      case InputType.Good => events.foreach(e => producer.produce(config.nsqGoodSinkTopicName, e._1.getBytes())) // To stdout
+      case InputType.Bad => events.foreach(e => producer.produce(config.nsqBadTopicName, e._1.getBytes())) // To stderr
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: we can refactor this in a method too

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/NSQSource.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  */
+class NSQSource(config: KinesisEnrichConfig, igluResolver: Resolver, enrichmentRegistry: EnrichmentRegistry, tracker: Option[Tracker])
+  extends AbstractSource(config, igluResolver, enrichmentRegistry, tracker) {
+
+  /**
+    * Consumer will be started to wait new message.
+    */
+  def run = {
+
+    class nsqCallback extends NSQMessageCallback {
```
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/NSQSource.scala]: you can refactor this as:

```scala
val nsqCallback  = new NSQMessageCallback {
  // ...
}
```

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala (hunk 6 above)
Comment (BenFradet) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/NSQSink.scala]: should we specify the charset in `getBytes()`

File: 2-collectors/scala-stream-collector/project/Dependencies.scala [hunk 8]
```diff
@@ -34,6 +34,7 @@ object Dependencies {
     val yodaTime         = "2.1"
     val yodaConvert      = "1.2"
     val kafka            = "0.10.1.0"
+    val nsqClient       = "1.0.0.RC4"
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/project/Dependencies.scala]: is there a release incoming? I'm uneasy at the idea of having an RC as dependency.

**Comment by alexanderdean** [comment]
Also this commit: 9e950c3d58cf84dacdbb6e78eb6c68fe0c1ce932 should be broken out into the associated tickets...

**Pull request closed without merging**